framework/math/ParallelVector/parallel_vector.h

```diff
@@ -0,0 +1,157 @@
+#ifndef PARALLEL_VECTOR_H
+#define PARALLEL_VECTOR_H
+
+#include "chi_runtime.h"
+
+#include <vector>
+#include <cstdint>
+#include <map>
+
+#include <mpi.h>
+
+
+namespace chi_math
+{
+
+enum class OperationType : short
+{
+  SET_VALUE = 1,
+  ADD_VALUE = 2
+};
+
+
+/**
+ * An implementation of a parallel vector.
+ */
+class ParallelVector
+{
+public:
+  using iterator = std::vector<double>::iterator;
+  using const_iterator = std::vector<double>::const_iterator;
+
+public:
+  /**
+   * Initialize a parallel vector with the given local and global sizes with
+   * the given communicator whose entries are set to zero.
+   */
+  ParallelVector(const uint64_t local_size,
+                 const uint64_t global_size,
+                 const MPI_Comm communicator = MPI_COMM_WORLD);
+
+  ParallelVector(const ParallelVector& other)
+    : local_size_(other.local_size_),
+      global_size_(other.global_size_),
+      location_id_(other.location_id_),
+      process_count_(other.process_count_),
+      values_(other.values_),
+      comm_(other.comm_)
+  {}
+
+  ParallelVector(ParallelVector&& other)
+    : local_size_(other.local_size_),
+      global_size_(other.global_size_),
+      location_id_(other.location_id_),
+      process_count_(other.process_count_),
+      values_(other.values_),
+      comm_(other.comm_)
+  {}
+
+  /// Return the size of the locally owned portion of the parallel vector.
+  uint64_t LocalSize() const { return local_size_; }
+
+  /// Return the global size of the parallel vector.
+  uint64_t GlobalSize() const { return global_size_;}
+
+  /**
+   * Read only accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  double operator[](const int64_t local_id) const;
+
+  /**
+   * Read/write accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor only allows access to the locally owned entries
+   *       of the parallel vector, and does not allow access for any data
+   *       beyond local_size_ that may exist in derived classes.
+   */
+  double& operator[](const int64_t local_id);
```

Comment:
Another problem here. If you use this function in pure read fashion, now all of a sudden you can't read ghosts just because you are using it in a method/function where it is non-const.